TEST_CASE_FIXTURE: A new fixture macro for allowing persistent fixtures throughout a TEST_CASE

KStocky opened the pull request:
## Description

This PR introduces a new `TEST_CASE` macro called `TEST_CASE_FIXTURE`. `TEST_CASE_FIXTURE` offers the same functionality as `TEST_CASE_METHOD` except for one difference. The object on which the test method is invoked is only created once for all invocations of the test case. The object is created just after the `testCaseStarting` event is broadcast and the object is destroyed just before the `testCaseEnding` event is broadcast.

The main motivation for this new functionality is to allow `TEST_CASE`s to do expensive setup and teardown once per `TEST_CASE`, without having to resort to abusing event listeners or static function variables with manual initialization.

<!--
Describe the what and the why of your pull request. Remember that these two
are usually a bit different. As an example, if you have made various changes
to decrease the number of new strings allocated, that's what. The why probably
was that you have a large set of tests and found that this speeds them up.
-->

## GitHub Issues

This PR hopes to be an implementation of #1602

KStocky committed:
Initial implementation of Test cases with persistent fixtures. With a simple example

KStocky committed:
Ignore all files with the name CMakeUserPresets.json

KStocky committed:
Adding acceptance tests for TEST_CASE_FIXTURE

KStocky committed:
Adding unapproved.txt files to gitignore

KStocky committed:
Make Fixture example clearer

KStocky committed:
Added documentation for TEST_CASE_FIXTURE as well as reorganised some of the existing documentation around the existing test fixture macros.

KStocky committed:
Ran updateDocumentToC.py to generate new tables of contents

KStocky committed:
Add version placeholder for TEST_CASE_FIXTURE macro

KStocky commented on docs/cmake-integration.md:
This was done when I was generating ToC after I had added my documentation to test-fixtures.md. I can remove it if this should be done in another PR since it isn't exactly to do with this PR at all.

horenmar replied:
This is fine.

KStocky commented on docs/other-macros.md:
I have removed this section and placed it in with the test-fixtures documentation since it is a macro for creating a test fixture.

KStocky commented on docs/test-fixtures.md:
I have reorganised this file to better document all of the fixture macros in one place and to compare and contrast them with each other.

KStocky commented on examples/CMakeLists.txt:
I couldn't find out what the naming scheme of examples is... Happy to rename this once I know what the name should be!

KStocky commented on src/catch2/catch_test_macros.hpp:
```diff
@@ -195,6 +198,7 @@
   #define TEST_CASE( ... )  INTERNAL_CATCH_TESTCASE_NO_REGISTRATION(INTERNAL_CATCH_UNIQUE_NAME( CATCH2_INTERNAL_TEST_ ), __VA_ARGS__)
   #define TEST_CASE_METHOD( className, ... ) INTERNAL_CATCH_TESTCASE_NO_REGISTRATION(INTERNAL_CATCH_UNIQUE_NAME( CATCH2_INTERNAL_TEST_ ))
   #define METHOD_AS_TEST_CASE( method, ... )
+  #define TEST_CASE_FIXTURE( className, ... )
```
I just copied what TEST_CASE_METHOD does in this case. It would seem that these are purposefully unimplemented

horenmar replied:
`TEST_CASE_METHOD` does this though

```cpp
  #define TEST_CASE_METHOD( className, ... ) INTERNAL_CATCH_TESTCASE_NO_REGISTRATION(INTERNAL_CATCH_UNIQUE_NAME( CATCH2_INTERNAL_TEST_ ))
```

horenmar replied:
Also please add the macro to `tests/ExtraTests/X02-DisabledMacros.cpp` to check that the compilation still works when this preprocessor branch is compiled.

KStocky replied:
Yup this was what was causing the majority of the CI failures. Fixed now!

KStocky commented on src/catch2/internal/catch_test_registry.hpp:
```diff
@@ -47,6 +47,33 @@ Detail::unique_ptr<ITestInvoker> makeTestInvoker( void (C::*testAsMethod)() ) {
+    }
+
+    void invoke() const override {
+        auto* f = const_cast<C*>( m_fixture.get() );
```
Just want to draw attention to this because const_cast is usually a sin and UB. However, this is a valid use-case of it that does not invoke UB since m_fixture does not point to a const object. This const_cast is required due to how Detail::unique_ptr implements get(). It differs from std::unique_ptr since it will return a pointer to const if the unique_ptr itself is const. This is not what std::unique_ptr does. https://en.cppreference.com/w/cpp/memory/unique_ptr/get.

The other option was to manually manage a raw pointer and add special member functions to ensure that it is handled correctly, but I have erred with const_cast in this case. Happy to change this though

horenmar replied:
The implementation difference is intentional, the shallow const on language pointers makes it annoyingly easy to throw away logical constness. So the const cast here is fine-ish.

Here is a thought though: why not make the fixture member function const?

It would be inconsistent with the `TEST_CASE_METHOD` macro, but this is fine; inconsistency can be okay if it brings improvement. The advantage is that it would be much harder to accidentally mutate the fixture -> this is much more important here than for `TEST_CASE_METHOD`, as the fixture here is persistent between test runs. It is still possible to allow fixture mutation, by marking the intended-to-change member variable with `mutable`.

I actually took your branch and tried a quick sketch of this and I think it is better, but I am willing to hear counter arguments.

KStocky replied:
I have zero issues with this honestly. I agree that this would be a better course of action and would make test writers  think about what they want to be mutatable. Const as the default is never usually a bad thing. I will make this change, and also note it in the docs, and perhaps also make a example to demonstrate it. Thanks for the great improvement!

KStocky commented on src/catch2/internal/catch_test_registry.hpp:
```diff
@@ -47,6 +47,33 @@ Detail::unique_ptr<ITestInvoker> makeTestInvoker( void (C::*testAsMethod)() ) {
+};
+
+template<typename C>
+Detail::unique_ptr<ITestInvoker> makeTestInvokerFixture( void ( C::*testAsMethod )() ) {
```
I could not make a `makeTestInvoker` overload since it would have the same function signature as the factory function for TEST_CASE_METHOD.

KStocky commented on src/catch2/catch_test_macros.hpp:
```diff
@@ -142,6 +144,7 @@
   #define TEST_CASE( ... ) INTERNAL_CATCH_TESTCASE( __VA_ARGS__ )
   #define TEST_CASE_METHOD( className, ... ) INTERNAL_CATCH_TEST_CASE_METHOD( className, __VA_ARGS__ )
   #define METHOD_AS_TEST_CASE( method, ... ) INTERNAL_CATCH_METHOD_AS_TEST_CASE( method, __VA_ARGS__ )
+  #define TEST_CASE_FIXTURE( className, ... ) INTERNAL_CATCH_TEST_CASE_FIXTURE( className, __VA_ARGS__ )
```
I am not tied to this name in any way. I think it is an OK name because this is the typical behaviour of a fixture in other testing frameworks i.e. the fixture objects is the same fixture object for all test cases that use it. 

I toyed with TEST_CASE_PERSISTENT_FIXTURE and TEST_CASE_WITH_SETUP among other names but they seemed to be a bit too long... In the end, I decided to just go with this name for now and if anyone has a better suggestion, I am more than happy to go with that.

horenmar replied:
> because this is the typical behaviour of a fixture in other testing frameworks i.e. the fixture objects is the same fixture object for all test cases that use it.

This is not true. The most used C++ test framework is GTest, which creates separate fixture instance per test. Catch2's current support for this does the same, and thus also Doctest and other clones. If I remember CppUnit correctly, it also creates a separate instance of fixture per test.

Now you might argue that none of these is a good match, because only Catch2 and its derivatives support defining multiple partial test cases in single test case body, and for GTest and CppUnit we are talking about different test case bodies, but I think of Catch2's test case trees as a short hand for defining multiple test cases, which would make them equivalent.

I am not saying that `TEST_CASE_FIXTURE` is the wrong name, just disputing the assertion about the typical behaviour of a (C++) test framework.

----------------

If I am reading the xUnit docs correctly, it also creates a separate fixture instance per test. Same for Python's unittest.

----------------

Personally I slightly prefer `TEST_CASE_PERSISTENT_FIXTURE`.

KStocky replied:
TEST_CASE_PERSISTENT_FIXTURE it is then!

KStocky commented on docs/test-fixtures.md:
```diff
@@ -30,8 +47,59 @@ class UniqueTestsFixture {
+
+### 3. `TEST_CASE_FIXTURE`
+
+> [Introduced](link-to-issue-or-PR) in Catch2 X.Y.Z
```
I think this is the correct way to introduce a version placeholder...

horenmar replied:
The suffix is correct (the release scripts will pick it up and update to the new version), but the target has to be done manually, so `https://github.com/catchorg/Catch2/pull/2885` instead of the `link-to-issue-or-PR`.

KStocky replied:
Yup this was something that should have been more obvious to me... My bad. And fixed now!

KStocky commented on docs/test-fixtures.md:
```diff
@@ -30,8 +47,59 @@ class UniqueTestsFixture {
-The two test cases here will create uniquely-named derived classes of UniqueTestsFixture and thus can access the `getID()` protected method and `conn` member variables. This ensures that both the test cases are able to create a DBConnection using the same method (DRY principle) and that any ID's created are unique such that the order that tests are executed does not matter.
+The two test cases here will create uniquely-named derived classes of UniqueTestsFixture and thus can access the `getID()` protected method and `conn` member variables. This ensures that both the test cases are able to create a DBConnection using the same method (DRY principle) and that any ID's created are unique such that the order that tests are executed does not matter. 
+
+### 2. `METHOD_AS_TEST_CASE`
```
Cut and pasted from other-macros.md

KStocky committed:
Remove unapproved.txt that was accidently committed

KStocky committed:
Merge pull request #2 from KStocky/AddCMakeUserPresetsToGitIgnore

Ignore all files with the name CMakeUserPresets.json

KStocky committed:
Merge pull request #3 from KStocky/AddUnapprovedFilesToGitIgnore

Add unapproved files to gitignore

KStocky committed:
Merge branch 'devel' into TestCaseWithSetup

horenmar commented on docs/test-fixtures.md:
```diff
@@ -1,9 +1,26 @@
-Although Catch allows you to group tests together as [sections within a test case](test-cases-and-sections.md), it can still be convenient, sometimes, to group them using a more traditional test fixture. Catch fully supports this too. You define the test fixture as a simple structure:
+## Non-Templated test fixtures
+
+Although Catch allows you to group tests together as [sections within a test case](test-cases-and-sections.md), it can still be convenient, sometimes, to group them using a more traditional test fixture. Catch fully supports this too with 3 different macros for non-templated test fixtures. They are: 
```
Formatting nitpick first: new docs use linebreaks at 80 chars (well, more like 72, you know how it is).

New docs also use Catch2 to refers to the project. (The reason I don't mass update this in the old docs is that it serves as a good heuristic on how old the docs are, and how they will be written).

KStocky replied:
I think I have addressed this now. However, I was not able to follow this rule with regard to tables in markdown because taking a line break would introduce a new row in the table... If you have any suggestions on how to solve this issue, I will be more than happy to do that :)

horenmar replied:
I don't think it is possible to do with tables in github's markdown, so it's fine to keep tables without linebreaks.

horenmar commented on docs/test-fixtures.md:
````diff
@@ -30,8 +47,59 @@ class UniqueTestsFixture {
+    }
+}
+```
+This test case will be executed twice as there are two leaf sections. On the first run `val` will be `0` and on the second run `val` will be `1`. This is useful if you would like to share some expensive setup code with all runs of your test case which can't be done at static initialization time.
````
I am not particularly happy about the example/explanation combo. My first reaction was "well, how's that different from using static variable here".

I can't give you better version right now, but I think this should be focused on the performance difference from `TEST_CASE_METHOD`, maybe have something like

```cpp
// Pretend doing hard work
std::this_thread::sleep_for(2s);
```

in the fixture constructor and explain that `TEST_CASE_FIXTURE` takes 2 seconds, while `TEST_CASE_METHOD` needs 2/4/6/8/... seconds to execute.

horenmar replied:
This should probably also explain that two different `TEST_CASE_FIXTURE`s have **separate** instances of fixture, i.e., the fixture is not static globally, just for the paths defined in one `TEST_CASE`.

KStocky replied:
I have redone the example and the documentation for this feature now. I feel like it now better highlights why a test writer would use this over TEST_CASE_METHOD.

horenmar commented:
As an FYI, I am (rather obviously) not finished with the review, but my bus ride is over and I'll have to return to this later.

horenmar commented on docs/test-fixtures.md:
````diff
@@ -30,8 +47,59 @@ class UniqueTestsFixture {
+`TEST_CASE_FIXTURE` behaves in the same way as [TEST_CASE_METHOD](#1-test_case_method) except that there will only be one instance created throughout the entire run of a test case. To demonstrate this have a look at the following example:
+
+```cpp
+struct MyFixture{
````
Also a formatting nitpick: space.
```suggestion
struct MyFixture {
```

KStocky replied:
Yup ran the formatter over this file now. I have also named it more appropriately.

horenmar committed:
constify the member function on fixtures

horenmar commented on src/catch2/catch_test_case_info.hpp:
```diff
@@ -112,6 +112,14 @@ namespace Catch {
         TestCaseHandle(TestCaseInfo* info, ITestInvoker* invoker) :
             m_info(info), m_invoker(invoker) {}
 
+        void testCaseStarting() const { 
```
Not a big fan of this name. It is consistent with the event names, but it does not describe the reason why it exists well. How about "prepareTestCase" instead?

(And same for `testCaseEnding` below.)

KStocky replied:
Yup as below, I have renamed these prepareTestCase and tearDownTestCase

horenmar commented on src/catch2/interfaces/catch_interfaces_test_invoker.hpp:
```diff
@@ -12,6 +12,8 @@ namespace Catch {
     class ITestInvoker {
     public:
+        virtual void testCaseStarting();
+        virtual void testCaseEnding();
```
Same note about name here.

KStocky replied:
I have named these `prepareTestCase` and `tearDownTestCase` now

horenmar reviewed (changes requested):
Overall I am convinced that the functionality is useful, but I want some changes before merging.

KStocky commented:
> Overall I am convinced that the functionality is useful, but I want some changes before merging.

Awesome. One other question, What is the naming scheme of the examples? There appears to be one but I could not figure it out. Hence why I just left my example as being called "Fixture.cpp".

KStocky committed:
Merge branch 'catchorg:devel' into TestCaseWithSetup

KStocky committed:
Renaming TEST_CASE_FIXTURE to TEST_CASE_PERSISTENT_FIXTURE. Also add licence to Fixture example

KStocky committed:
Rename the example file
Fix the license on the file
Improve the example to make the use cases much clearer. Also fix up the documentation

KStocky committed:
Ensured that TEST_CASE_PERSISTENT_FIXTURE works with CATCH_CONFIG_DISABLE

KStocky committed:
Rename testCaseStarting to prepareTestCase and renamed testCaseEnding to tearDownTestCase

KStocky committed:
Further clarifying the difference between TEST_CASE_METHOD and TEST_CASE_PERSISTENT_FIXTURE. Also adding the example to the list of examples

horenmar commented on examples/111-Fix-PersistentFixture.cpp:
```diff
@@ -0,0 +1,74 @@
+
+// Fixture.cpp
+
+// Catch has three ways to express fixtures:
```
```suggestion
// Catch2 has three ways to express fixtures:
```

horenmar committed:
Update examples/111-Fix-PersistentFixture.cpp

horenmar commented:
Thanks
